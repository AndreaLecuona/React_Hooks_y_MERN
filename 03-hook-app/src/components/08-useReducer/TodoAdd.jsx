import React from 'react'
import { useForm } from '../../hooks/useForm'

export const TodoAdd = ( { handleAdd } ) => {

    const [ { description }, handleInputChange, reset ] = useForm({
        description: ''
    })

    const handleSubmit = (e) => {
        e.preventDefault();

        if( description.trim().length <= 1 ) {
            return
        };

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        //reemplazo este dispatch con handle add para que el dispatch se haga en el componente padre
        handleAdd( newTodo );
        reset();
    }

    return (
        <div className="form-group">
            <form onSubmit={ handleSubmit }>
                <input type="text"
                    name="description"
                    className="form-control"
                    placeholder="Tarea por realizar..."
                    autoComplete="off"
                    value={ description }
                    onChange={ handleInputChange }
                />

                <button type="submit"
                    className="btn btn-outline-primary personal-btn mt-3">
                    Agregar a la lista
                </button>
            </form>
        </div>
    )
}
