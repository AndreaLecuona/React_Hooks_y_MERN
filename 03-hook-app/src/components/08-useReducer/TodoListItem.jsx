import React from 'react'

export const TodoListItem = ( { todo, index, handleDelete, handleToggle } ) => {
    return (
        <li
            key={ todo.id }
            className="list-item"
        >
            <p 
                className={`${ todo.done && 'complete' }`}
                onClick={() => handleToggle( todo.id )}>

                    {index + 1}. { todo.desc }
            </p>

            <button 
                className="btn btn-danger personal-btn"
                onClick={() => handleDelete( todo.id )}>

                    Borrar
            </button>
        </li>
    )
}
