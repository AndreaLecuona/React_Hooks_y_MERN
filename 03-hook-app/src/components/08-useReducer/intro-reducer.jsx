const initialState = [{
    id: 1,
    todo: 'Comprar cafe',
    done: false
}];

//El use reducer centraliza todas las acciones posibles de la aplicacion, que van a afectar el state. No puede ser asíncrona, llamar al local storage, ni llamar a otras funciones desde su cuerpo o efectos secundarios (es pura)
//Recibe como argumentos un nuevo state y una accion(que a su vez puede o no tener un argumento) que va a modificar el initial state, y SIEMPRE devuelve un state. NUNCA NULL O UNDEFINED
const todoReducer = ( state = initialState, action ) => {

    if(action?.type === 'agregar nuevo item'){ //optional chaining: 'si action no es undefined o null, comparar con el string'
        return [ ...state, action.payload ];
    }

    return state;
};

let todos = todoReducer();


const newtodo = {
    id: 2,
    todo: 'Cocinar brownies',
    done: false
};

//Como .push() no se utiliza en react porque muta el objeto/modifica el state y rompe la app, tenemos que pasar el nuevo item de otra forma, junto con la acción
//en este objeto 'accion' que se enviará como argumento al reducer sse incluyen el nombre de la accion propiamente dicha y el objeto que involucra
//type y payload son nombres estandard, podrían ser los que quisiéramos
const agregarTodoAction = {
    type: 'agregar nuevo item',
    payload: newtodo
}


todos = todoReducer( todos, agregarTodoAction );


console.log(todos);