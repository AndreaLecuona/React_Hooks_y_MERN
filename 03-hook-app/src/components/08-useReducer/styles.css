h1{
    margin-left: 30px;
}
ul{
    list-style: none;
}
li{
    display: flex;
    justify-content: space-between;
    border-bottom: 0.5px solid black;
    padding: 2rem 0;
}
.form-group{
    width: 50%;
    margin-left: 30px;
}
