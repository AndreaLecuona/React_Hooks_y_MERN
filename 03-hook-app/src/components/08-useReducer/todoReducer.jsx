
export const todoReducer = (state = [], action) => {
    
    switch ( action.type ) {
        //debe conocer de antemano todas las posibles 'acciones' que le mandemos
        case 'add':
            return [...state, action.payload];

        case 'delete':
            return state.filter( todo => todo.id !== action.payload); //action-payload toma el id

        case 'toggle':
            return state.map( todo => 
                ( todo.id === action.payload ) ? { ...todo, done: !todo.done } : todo
            );

        //VERSION LARGA DE TOGGLE
        // case 'toggle':
        //     return state.map( todo => {
        //         if( todo.id === action.payload ){
        //             return {
        //                 ...todo,
        //                 done: !todo.done
        //             }
        //         } else {
        //             return todo;
        //         }
        //     })

        default:
            return state; //para asegurarnos que devuelva un state
    }

};