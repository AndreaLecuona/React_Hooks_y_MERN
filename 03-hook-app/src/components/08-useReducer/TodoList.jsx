import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ( {todos, handleDelete, handleToggle} ) => {
    return (
        <ul className="list-movies">
            {
                todos.map((todo, i) =>

                    <TodoListItem
                        key ={ todo.id }
                        todo ={ todo }
                        index ={ i }
                        handleDelete = { handleDelete }
                        handleToggle = { handleToggle }
                    />

                )
            }
        </ul>
    )
}
