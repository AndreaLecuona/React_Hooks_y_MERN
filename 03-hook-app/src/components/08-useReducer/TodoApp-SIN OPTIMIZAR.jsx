import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
import { useForm } from '../../hooks/useForm';

import './styles.css';

//const [state, dispatch] = useReducer(reducer, initialState, init)
//init: Es una funcion para inicializar el state en caso que Ã©ste sea procesado o haga varias acciones (retorna el initial state)
//dispatch: Ayuda a "disparar"/enviar las acciones hacia el reducer para volver a renderizar en caso de haberse producido un cambio en el state


// const initialState = [{
//     id: new Date().getTime(),
//     desc: "Barry Lyndon - Stanley Kubrick",
//     done: false
// }];

const init = () => {
    //LEER DE LOCAL STORAGE Y MANDARLO COMO INITIAL STATE
    return JSON.parse(localStorage.getItem( 'todos' )) || [];

    // return(
    //     [{
    //         id: new Date().getTime(), //simulacion de id
    //         desc: "Barry Lyndon - Stanley Kubrick",
    //         done: false
    //     }]
    // )
}


export const TodoApp = () => {
    
    //MANEJO DE INPUT CON CUSTOM HOOK    const [ formValues, handleInputChange] = useForm({...
    const [ { description }, handleInputChange, reset ] = useForm({
        description: ''
    })

    //Extraigo del resultado del reducer una lista de items
    const [ todos, dispatch ] = useReducer(todoReducer, [], init);
    console.log( description );


    //GUARDAR EN LOCAL STORAGE
    //guardar en primer renderizado y cada vez que la lista cambie
    useEffect(() => {
        localStorage.setItem( 'todos', JSON.stringify( todos ) );
    }, [todos]);

    const handleDelete = (todoId) => {
        console.log(todoId);

        const action = {
            type: 'delete',
            payload: todoId
        };

        dispatch( action );
    }

    const handleToggle = (todoId) => {
        dispatch( {
            type: 'toggle',
            payload: todoId
        })
    }


    //MANEJO DE SUBMIT
    const handleSubmit = (e) => {
        e.preventDefault();

        if( description.trim().length <= 1 ) {
            return
        };

        //1. creo el nuevo item
        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        //2. lo agrego a una accion y defino el tipo de accion
        const action = {
            type: 'add',
            payload: newTodo
        };

        //3. mando la accion al reducer con dispatch
        dispatch( action );
        reset();
    }

    return (
        <div>
            <h1>Pelis para ver en cuarentena ( {todos.length} )</h1>
            <hr/>

            <div className="row">

                <div className="col-7">
                    
                    <ul className="list-group list-group-flush">
                        {
                            todos.map((todo, i) =>
                                {return <li
                                    key={todo.id}
                                    className="list-group-item"
                                >
                                    <p className={ `${ todo.done && 'complete' }` } 
                                    onClick={ () => handleToggle(todo.id) 
                                    }>
                                        {i + 1}. {todo.desc}
                                    </p>
                                    <button className="btn btn-danger"
                                    onClick={() => handleDelete(todo.id)}
                                    >
                                        Borrar
                                    </button>
                                </li>}
                            )
                        }
                    </ul>

                </div>

                <div className="col-5">
                    
                    <h4>Me falta ver...</h4>
                    <hr/>

                    <form onSubmit={handleSubmit}>
                        <input type="text"
                            name="description"
                            className="form-control"
                            placeholder="peli - director"
                            autoComplete="off"
                            value={ description }
                            onChange={ handleInputChange }
                        />

                        <button type="submit"
                            className="btn btn-outline-primary mt-2">
                            Agregar
                        </button>
                    </form>

                </div>

            </div>
            
        </div>
    )
}
