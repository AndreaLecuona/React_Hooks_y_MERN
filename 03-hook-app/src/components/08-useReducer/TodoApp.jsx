import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
// import { useForm } from '../../hooks/useForm';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

import './styles.css';

//const [state, dispatch] = useReducer(reducer, initialState, init)
//init: Es una funcion para inicializar el state en caso que Ã©ste sea procesado o haga varias acciones (retorna el initial state)
//dispatch: Ayuda a "disparar"/enviar las acciones hacia el reducer para volver a renderizar en caso de haberse producido un cambio en el state


// const initialState = [{
//     id: new Date().getTime(),
//     desc: "Barry Lyndon - Stanley Kubrick",
//     done: false
// }];

const init = () => {
    //LEER DE LOCAL STORAGE Y MANDARLO COMO INITIAL STATE
    return JSON.parse(localStorage.getItem( 'todos' )) || [];

    // return(
    //     [{
    //         id: new Date().getTime(), //simulacion de id
    //         desc: "Barry Lyndon - Stanley Kubrick",
    //         done: false
    //     }]
    // )
}


export const TodoApp = () => {

    //Extraigo del resultado del reducer una lista de items
    const [ todos, dispatch ] = useReducer(todoReducer, [], init);

    //GUARDAR EN LOCAL STORAGE
    //guardar en primer renderizado y cada vez que la lista cambie
    useEffect(() => {
        localStorage.setItem( 'todos', JSON.stringify( todos ) );
    }, [todos]);

    const handleDelete = (todoId) => {
        console.log(todoId);

        const action = {
            type: 'delete',
            payload: todoId
        };

        dispatch( action );
    }

    const handleToggle = (todoId) => {
        dispatch( {
            type: 'toggle',
            payload: todoId
        })
    }

    const handleAdd = ( newTodo ) => {

        dispatch({
            type: 'add',
            payload: newTodo
        });
    }

    return (
        <div className="wrapper my-5">

            <div className="row both-sides mx-5">
                <h1>TodoApp ({ todos.length })</h1>

                <div className="col-sm-12 col-lg-5 mt-5">
                    <TodoList
                        todos={ todos }
                        handleDelete={ handleDelete }
                        handleToggle={ handleToggle }
                    />

                </div>

                <div className="col-sm-12 col-lg-7 mt-5">            
                    <TodoAdd
                    handleAdd={ handleAdd }
                    />

                </div>

            </div>
            
        </div>
    )
}
