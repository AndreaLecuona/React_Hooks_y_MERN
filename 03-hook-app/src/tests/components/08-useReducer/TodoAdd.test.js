import React from 'react';
import { shallow } from "enzyme";
import { TodoAdd } from "../../../components/08-useReducer/TodoAdd";

describe('Pruebas en TodoAdd', () => {

    const handleAdd = jest.fn();

    const wrapper = shallow( 
        <TodoAdd
            handleAdd={ handleAdd }
        /> 
    );

    test('debe renderizarse correctamente', () => {
        
        expect( wrapper ).toMatchSnapshot();

    })

    test('cuando NO paso un valor a descripcion, el submit formulario no debe llegar a llamar a handleAdd', () => {
        //si no le paso valor a desc, entonces al hacer submit, se pone en juego el e.preventdefault y la validacion de campo vacío,
        //  lo cual hace que se detenga el submit y no se llegue a llamar a handleAdd dandole un argumento desc vacío

        //puede hacerse simulando el submit (como ya hicimos en otra prueba)
        //o puede hacerse referenciando al submit como propiedad:
        const formSubmit = wrapper.find( 'form' ).prop( 'onSubmit' );

        //si chequeamos en consola, vemos que esta constante se considera una función
        //console.log(formSubmit); //devuelve [Function: handleSubmit]

        //podemos hacer uso entonces, de la constante como función, pero necesitamos pasarle como parámetro la función preventDefault para que la use
        // formSubmit sin argumentos crashea
        // preventDeafult con la desc como objeto vacío, es lo que necesitamos para testear
        formSubmit( { preventDefault(){} } );


        //esperamos que NO sea llamado
        expect( handleAdd ).toHaveBeenCalledTimes( 0 );

    })
    
    test('debe llamar a handleAdd cuando hay un argumento válido', () => {

        //modificamos el valor del input para que tenga algo y no esté vacío
        const value = 'Limpiar...';

        wrapper.find( 'input' ).simulate( 'change', {
            target: {
                value,
                name: 'description'
            }
        } )

        //capturamos y llamamos a la simulacion del evento submit
        const formSubmit = wrapper.find( 'form' ).prop( 'onSubmit' );

        formSubmit( { preventDefault(){} } );

        //esperamos que handleAdd haya sido llamado 1 vez
        expect( handleAdd ).toHaveBeenCalledTimes( 1 );

        //esperamos que haya sido llamado con un objeto como argumento:
        //expect( handleAdd ).toHaveBeenCalledWith( expect.any( Object ) ); //no puedo escribirlo así porque un objeto vacío también pasaría
        expect( handleAdd ).toHaveBeenCalledWith( {
            id: expect.any( Number ), //le decimos a jest que no importa el numero que salga acá mientras sea de tipo número (porque en la app esta armado para que siempre sea random, y si testeamos con numero especifico, nos va a hacer fallar las pruebas)
            desc: value,
            done: false
        } )

        //el submit tambien debe llamar el reset() luego de llamar a handleAdd
        expect( wrapper.find( 'input' ).prop( 'value' ) ).toBe( '' );
    })
      
})
