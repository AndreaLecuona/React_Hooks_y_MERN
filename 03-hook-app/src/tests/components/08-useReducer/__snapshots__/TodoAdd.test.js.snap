// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en TodoAdd debe renderizarse correctamente 1`] = `
<div
  className="form-group"
>
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Nombre de la pelÃ­cula"
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary personal-btn mt-3"
      type="submit"
    >
      Agregar a la lista
    </button>
  </form>
</div>
`;
